src/components/tl/ucp/allgather/allgather_sparbit.c

```diff
@@ -0,0 +1,156 @@
+/**
+ * Copyright (c) 2021-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * See file LICENSE for terms.
+ */
+#include "config.h"
+#include "tl_ucp.h"
+#include "allgather.h"
+#include "core/ucc_progress_queue.h"
+#include "tl_ucp_sendrecv.h"
+#include "utils/ucc_math.h"
+#include "utils/ucc_coll_utils.h"
+#include "components/mc/ucc_mc.h"
+#include <stdio.h>
+
+ucc_status_t ucc_tl_ucp_allgather_sparbit_init(ucc_base_coll_args_t *coll_args,
+                                               ucc_base_team_t      *team,
+                                               ucc_coll_task_t     **task_h)
+{
+    ucc_tl_ucp_task_t *task   = ucc_tl_ucp_init_task(coll_args, team);
+    ucc_status_t       status = UCC_OK;
+
+    if (!ucc_coll_args_is_predefined_dt(&TASK_ARGS(task), UCC_RANK_INVALID)) {
+        tl_error(UCC_TASK_LIB(task), "user defined datatype is not supported");
+        status = UCC_ERR_NOT_SUPPORTED;
+        goto out;
+    }
+    tl_trace(UCC_TASK_LIB(task), "ucc_tl_ucp_allgather_sparbit_init");
+
+    task->super.post          = ucc_tl_ucp_allgather_sparbit_start;
+    task->super.progress      = ucc_tl_ucp_allgather_sparbit_progress;
+    task->allgather_sparbit.i = 0; // setup iteration
+    task->allgather_sparbit.data_expected = 1;
+out:
+    if (status != UCC_OK) {
+        ucc_tl_ucp_put_task(task);
+        return status;
+    }
+
+    *task_h = &task->super;
+    return status;
+}
+
+// return nearest greater log pow2
+static inline uint32_t highest_log2(uint32_t n)
+{
+    int x = __builtin_clz(n); // leading zeros
+    if (!(n & (n - 1)))
+        return 31 - x; // pow2 case
+    return 31 - x + 1;
+}
+
+/* Inspired by implementation: https://github.com/open-mpi/ompi/blob/main/ompi/mca/coll/base/coll_base_allgather.c */
+void ucc_tl_ucp_allgather_sparbit_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task      = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team      = TASK_TEAM(task);
+    ucc_rank_t         trank     = UCC_TL_TEAM_RANK(team);
+    ucc_rank_t         tsize     = UCC_TL_TEAM_SIZE(team);
+    void              *rbuf      = TASK_ARGS(task).dst.info.buffer;
+    ucc_memory_type_t  rmem      = TASK_ARGS(task).dst.info.mem_type;
+    ucc_datatype_t     dt        = TASK_ARGS(task).dst.info.datatype;
+    size_t             count     = TASK_ARGS(task).dst.info.count;
+    size_t             data_size = (count / tsize) * ucc_dt_size(dt);
+    uint32_t   i = task->allgather_sparbit.i; // restore iteration number
+    ucc_rank_t recvfrom, sendto;
+    size_t     distance;
+    uint32_t   last_ignore, ignore_steps, data_expected, transfer_count;
+    uint32_t   tsize_log, exclusion;
+    void      *tmprecv, *tmpsend;
+
+    // here we can't made any progress while transfers from previous step are running, emulation of wait all in async manier
```

Comment:
```suggestion
    // here we can't make any progress while transfers from previous step are running, emulation of wait all in async manner
```
I am not sure to understand the word "emulation" here

Reply:
It is mostly about that this is re-enter function and we must to wait all ops from previous step, in algo description and open mpi implementation wait all is used